webkit2/jscore.jsc-value.lisp

```diff
@@ -164,3 +164,115 @@
+  "The Lisp data type used when translating objects from JavaScript to Lisp.")
+(export '*js-object-type*)
+
+(declaim (ftype (function (t &key (:null-value t)
+                             (:undefined-value t)
+                             (:false-value t)
+                             (:true-value t)
+                             (:function-value t)
+                             (:array-type (or null (member :list :vector)))
+                             (:object-type (or null (member :alist :plist :hash-table)))))
+                jsc-value-to-lisp))
+(defun jsc-value-to-lisp (jsc-value &key (null-value *js-null-value*)
+                                      (undefined-value *js-undefined-value*)
+                                      (false-value *js-false-value*)
+                                      (true-value *js-true-value*)
+                                      (function-value *js-function-value*)
+                                      (array-type *js-array-type*)
+                                      (object-type *js-object-type*))
+  "Translate a JSC-VALUE to a lisp value.
+Translates:
+- JS strings to strings.
+- JS numbers to floating numbers.
+- true to TRUE-VALUE (t by default). Also see `*js-true-value*'.
+- false to FALSE-VALUE (nil by default). Also see `*js-false-value*'.
+- null to NULL-VALUE (:null by default). Also see `*js-null-value*'.
+- undefined to UNDEFINED-VALUE (:undefined by default). Also see `*js-undefined-value*'.
+- JS functions to FUNCTION-VALUE (:function by default). This will change in the future!
+  Also see `*js-function-value*'.
+- JS arrays to either lists or vectors (:list and :vector ARRAY-TYPE respectively).
+  Also see `*js-array-type*'.
+- JS objects (also see `*js-object-type*') to:
+  - hash-tables (if OBJECT-TYPE is :hash-table, default) with string keys,
+  - alists (if OBJECT-TYPE is :alist) with string keys,
+  - plists (if OBJECT-TYPE is :plist) with keyword keys,"
+  (flet ((drain-properties (jsc-value array-p)
+           (loop with property-names = (jsc-value-object-enumerate-properties jsc-value)
+                 for index from 0
+                 until (cffi:null-pointer-p property-names)
+                 for property-name = (cffi:mem-aref property-names '(:pointer (:pointer :char)) index)
+                 until (cffi:null-pointer-p property-name)
+                 collect (let ((property-value
+                                 (jsc-value-to-lisp
+                                  (if array-p
+                                      (jsc-value-object-get-property-at-index jsc-value index)
+                                      (jsc-value-object-get-property jsc-value property-name)))))
+                           (if array-p
+                               property-value
+                               (cons (cffi:foreign-string-to-lisp property-name) property-value)))
+                   into properties
+                 finally (return
+                           (prog1
```

Comment:
Could just be a `progn` with `properties last, no?

Reply:
Fair. I've just liked `prog1` more there :)

I've changed it to `progn` in 8b8e221b90a2fc57916d96d4954a4e3e03253c18.